readme.md
# SemVerq

Semverq is a cli utility for:

- Validating semver format
- Accessing semver structures
- Convert semver to json

# Features

## Validation

For `-i` option, semverq validates the input string as a semver.

```
$ semverq -i '1.2.3-beta+dev-armhf'
(nothing is output)
$ echo $?
0
```

```
$ semverq -i '1.2.3-beta+dev+armhf'
invalid format as semver: 1.2.3-beta+dev+armhf
$ echo $?
150
```

If the `-i` option is not specified, semverq reads string from standard input.

```
$ semverq <<<'1.2.3-input+stdin'
$ echo $?
0
```

## Convert to json

For `-j` option, semverq converts the input to json.

```
$ semverq -j -i '1.2.3-beta+36a1d2f'
{
  "major": 1,
  "minor": 2,
  "patch": 3,
  "pre-release": "beta",
  "build": "36a1d2f"
}
```

If `pre-release` or `build` is not exists, the values of their fields will be `null`.

```
$ semverq -j -i '1.2.3'
{
  "major": 1,
  "minor": 2,
  "patch": 3,
  "pre-release": null,
  "build": null
}
```

## Accessing to parts

For `-q` option, access semver parts.

Get only major version:

```
$ semverq -q '.major' -i '1.2.3-beta+36a1d2f'
1
```

Build a json object with `pre-release` and `build`:

```
$ semverq -q '{ "pre-release": ".pre-release", "build": ".build" }' -i '1.2.3-beta+36a1d2f'
{ "pre-release": "beta", "build": "36a1d2f" }
```

### Accessor

Supported accessors are below:

- `.major`
  Major version.
- `.minor`
  Minor version.
- `.patch`
  Patch version.
- `.pre-release` or `.pre`
  pre-release version.
- `.build`
  Build metadata.
- `.version-core`
  Shorthand for `.major..minor..patch`.

## Checking for a match

For `-m` option, semverq checking for a match to given requirement.
The requirement is constructed from comma separated version constraint.
This matching syntax and rules are same as in [Rust's package system](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html).


Matching case:

```
$ semverq -m '>1.2.1, <=2.0.0' -i '1.2.3'
$ echo $?
0
```

Not matching case:

```
$ semverq -m '^1.2.1, 1.3.*' -i '1.2.3-beta+dev-armhf'
Version (1.2.3-beta+dev-armhf) does not match that requirement (^1.2.1, 1.3.*).
$ echo $?
152
```

Invalid requirement:

```
$ semverq -m '>1.2.1, <=2.0.0.' -i '1.2.3'
invalid format as version requirement: >1.2.1, <=2.0.0.
$ echo $?
151
```

# Packaging

For packaging semverq as a deb package:

```sh
docker-compose -f compose.yml run debbuilder
```




debian/changelog
semverq (0.1.0) unstable; urgency=low

  * Initial Release

 -- eldesh <nephits@gmail.com>  Wed, 25 May 2022 14:05:25 +0000


src/main.rs
use std::io::{self, BufRead};

use anyhow::{Context, Result};
use clap::{Arg, Command};
use serde_json::{self as json, json};

mod error;
use crate::error::Error;

fn version_core(version: &semver::Version) -> String {
    format!("{}.{}.{}", version.major, version.minor, version.patch)
}

fn process_query(version: &semver::Version, query: &str) -> String {
    query
        .replace(".major", &format!("{}", version.major))
        .replace(".minor", &format!("{}", version.minor))
        .replace(".patch", &format!("{}", version.patch))
        .replace(".version-core", &version_core(version))
        .replace(".pre-release", &format!("{}", version.pre))
        .replace(".pre", &format!("{}", version.pre))
        .replace(".build", &format!("{}", version.build))
}

fn cli(prog: &str) -> Command<'static> {
    Command::new(prog)
        .version(env!("CARGO_PKG_VERSION"))
        .about("A cli utility for validating semver, accessing semver structure and converting to json.")
        .arg(
            Arg::new("input")
                .short('i')
                .takes_value(true)
                .help("version string"),
        )
        .arg(
            Arg::new("to-json")
                .short('j')
                .help("convert the input string to a json"),
        )
        .arg(
            Arg::new("query")
                .short('q')
                .takes_value(true)
                .conflicts_with("to-json")
                .help("expand accessors"),
        )
        .arg(
            Arg::new("match")
                .short('m')
                .takes_value(true)
                .conflicts_with("to-json")
                .conflicts_with("query")
                .help("checking for a match"),
        )
}

fn main() {
    let m = cli("semverq").get_matches();

    let code = match process(m) {
        Ok(()) => 0,
        Err(err) => {
            if let Some(err) = err.downcast_ref::<Error>() {
                use Error::*;
                eprintln!("{}", err);
                match err {
                    InvalidSemver { .. } => 150,
                    InvalidVersionReq { .. } => 151,
                    NotMatchReq { .. } => 152,
                }
            } else {
                1
            }
        }
    };
    std::process::exit(code)
}

fn parse_semver(text: &str) -> Result<semver::Version> {
    Ok(
        semver::Version::parse(text).map_err(|source| Error::InvalidSemver {
            input: text.to_string(),
            source,
        })?,
    )
}

fn parse_semver_req(text: &str) -> Result<semver::VersionReq> {
    Ok(
        semver::VersionReq::parse(text).map_err(|source| Error::InvalidVersionReq {
            input: text.to_string(),
            source,
        })?,
    )
}

fn process(m: clap::ArgMatches) -> Result<()> {
    let version = if let Some(input) = m.value_of("input") {
        parse_semver(input)?
    } else {
        let stdin = io::stdin();
        let mut stdin = stdin.lock();
        let bytes = stdin
            .fill_buf()
            .with_context(|| "Failed to get the contents from the inner buffer of the stdin.")?;

        parse_semver(String::from_utf8_lossy(bytes).trim_end())?
    };

    if m.is_present("to-json") {
        println!(
            "{}",
            json::to_string_pretty(&json!({
                "major": version.major,
                "minor": version.minor,
                "patch": version.patch,
                "pre-release": if version.pre.is_empty() { json!(null) } else { json!(version.pre.as_str()) },
                "build": if version.build.is_empty() { json!(null) } else { json!(version.build.as_str()) }
            }))
            .unwrap()
        );
    } else if let Some(query) = m.value_of("query") {
        println!("{}", process_query(&version, query));
    } else if let Some(match_str) = m.value_of("match") {
        let req = parse_semver_req(match_str)?;
        if !req.matches(&version) {
            return Err(Error::NotMatchReq { version, req }.into());
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn query() {
        use serde_json::Value;
        let query = r#"{ "pre-release": ".pre-release", "build": ".build" }"#;
        let input = semver::Version::parse("1.2.3-beta+36a1d2f").unwrap();
        assert_eq!(
            serde_json::from_str::<Value>(&process_query(&input, query)).unwrap(),
            serde_json::from_str::<Value>(r#"{ "pre-release": "beta", "build": "36a1d2f" }"#)
                .unwrap()
        )
    }
}
